src/fhir_search/uri_parser.clj

```diff
@@ -0,0 +1,88 @@
+(ns fhir-search.uri-parser 
+  (:require [clojure.string :as str])
+  (:require [clojure.walk :refer [postwalk]])
+  (:import [java.net URI]))
+
+
+
+(defn url-encoding [url]
+  (let [uri (URI. url)
+        scheme (.getScheme uri)
+        authority (.getAuthority uri)
+        path (.getPath uri)
+        query (.getQuery uri)
+        fragment (.getFragment uri)]
+    (when uri
+      (str (when scheme (str scheme "://"))
+           authority path
+           (when query (str "?" query))
+           (when fragment (str "#" fragment))))))
+
+(defn values [query]
+  (->> 
+   (str/split query #"&")
+   (reduce (fn [result element]
+             (let [group (str/split element #"=") 
+                   first-group (first group) 
+                   second-group (second group)]
+               (conj result {:name (first (str/split first-group #":"))
+                             :modifiers (when (second (str/split first-group #":")) 
+                                          (keyword "fhir.search.modifier" (second (str/split first-group #":"))))
+                             :value second-group}))) [])))
+
+(defn path [string]
+  (let [path (remove empty? (-> (.getPath (URI. string)) (str/split #"/"))) 
+        structure {:type (last path)}]
+    (cond-> structure
+      (= 3 (count path)) (assoc :compartment {:type (first path) :id (second path)})
+      (= 2 (count path)) (assoc :component {:id (first path)}))))
+
+(defn params [string]
+  (let [query (.getQuery (URI. string))]
+    (when query
+      (hash-map :params {:join :fhir.search.join/and
+                                :values (values query)}))))
+
+(defn clean [m]
```

Comment:
lets move this fn outside this ns, because it is a general purpose fn. We can name it `fhir-search.complex` because it is about cleaning complex aggregations of values.